### DISCUSSION TITLE
Model's Token after training

### DISCUSSION THREAD

**Discussion by Trauter1** [d8ahazard/sd_dreambooth_extension#181] [General]
What's the token of the trained model if I want to use it to generate things?
```
[
    {
        "instance_prompt":      "1girl Arzaxi",
        "class_prompt":         "1girl",
        "instance_data_dir":    "../../../data/Arzaxi",
        "class_data_dir":       "../../../data/1girl"
    },
]
``` 
Is it Instance_prompt? But then if we use filewords, is the token still the same?
```
[
    {
        "instance_prompt":      "1girl Arzaxi, [filewords]",
        "class_prompt":         "1girl",
        "instance_data_dir":    "../../../data/Arzaxi",
        "class_data_dir":       "../../../data/1girl"
    },
]
```
![chrome_GXZMCi8HFR](https://user-images.githubusercontent.com/83187970/201715527-37e06639-9ab7-4075-a9a6-cf3fe1275c5b.png)
And using those settings, is there even a token?
All of the models I trained have heavy amount of bleeding so using token wasn't necessary but I want to eliminate it.
Sorry if im asking stupid question, but I looked online for information about it and can't find definite answer.

**Comment by d8ahazard** [comment]
The filewords bit is all @sgsdxzy, so I'll defer to them to answer that. ;)

**Comment by d8ahazard** [comment]
Also, moving to discussions as this is less an issue and more "How does this work?"

**Comment by sgsdxzy** [comment]
Just think it as you were doing textual inversion, and the addition of [filewords] is just like using textual_inversion_templates/subject_filewords.txt instead of textual_inversion_templates/subject.txt. In both cases, you can invoke your subject just by Arzaxi.

**Reply by sgsdxzy** [reply]
Also my guess about class prompt [filewords]:

> Class image can also read filewords, but in fact I know nothing about SD, torch or even machine learning in the first place, so I cannot say for sure. But as an instinct, if you want to train 'dog zxw, long fur, short tail, short leg', you should be using class images generated by 'dog, long fur, short tail, short leg'. And when invoking you should also write 'dog zxw, long fur, short tail, short leg', not just zxw, but the whole purpose of this is that you can also write 'dog zxw, long fur, long tail, long leg' to view your extended dog. I think that is what regularization trying to achieve, to make prompt composable. 
On the other hand if you train with 'dog zwx' and class image 'dog', you can use just zwx to invoke but it would be very hard to get a long tail version of zwx.
So probably we should implement the feature like this:
If user gives a class prompt without [filewords], keep it as is;
If user gives a class prompt with [filewords], and a directory with enough number of images, just read the filewords of the class image to be part of the class prompt.
If user gives a class prompt with [filewords], and an empty directory, and a nonzero class image number, we generate that many images, each time with a randomly chosen(or round-robin?) filewords of an instance image. Tag shuffle and dropout from settings also apply.

**Reply by Trauter1** [reply]
> Just think it as you were doing textual inversion, and the addition of [filewords] is just like using textual_inversion_templates/subject_filewords.txt instead of textual_inversion_templates/subject.txt. In both cases, you can invoke your subject just by Arzaxi.

Thank you for answering! 
I have one more thing to ask, if we tick this option
![chrome_OdfDPVH3Vg](https://user-images.githubusercontent.com/83187970/201721608-347c69de-7eaa-4626-94aa-b42768b3b947.png)
will our json file look like this? `"instance_prompt":      "[filewords]"`
If so, will typing this in instance prompt box
![chrome_CiMafUkIwc](https://user-images.githubusercontent.com/83187970/201721906-42b2abc4-dc5d-4d88-82f2-bc41329f834e.png)
Transform it into `"instance_prompt":      "1girl, [filewords]"` ?

**Reply by sgsdxzy** [reply]
> > Just think it as you were doing textual inversion, and the addition of [filewords] is just like using textual_inversion_templates/subject_filewords.txt instead of textual_inversion_templates/subject.txt. In both cases, you can invoke your subject just by Arzaxi.
> 
> Thank you for answering! I have one more thing to ask, if we tick this option ![chrome_OdfDPVH3Vg](https://user-images.githubusercontent.com/83187970/201721608-347c69de-7eaa-4626-94aa-b42768b3b947.png) will our json file look like this? `"instance_prompt": "[filewords]"` If so, will typing this in instance prompt box ![chrome_CiMafUkIwc](https://user-images.githubusercontent.com/83187970/201721906-42b2abc4-dc5d-4d88-82f2-bc41329f834e.png) Transform it into `"instance_prompt": "1girl, [filewords]"` ?

No, it's from another pr and totally override [filewords] behavior, your prompt is just ignored.

**Reply by Trauter1** [reply]
Ok, so I just put 1girl Arzaxi, [filewords] in instance prompt, it should automatically take tags from .txt file with the same name as image, without turning on this option right?
![chrome_cn1cnONLhQ](https://user-images.githubusercontent.com/83187970/201723039-4ad821d0-51e5-49c4-bc88-6005c7daef6d.png)

**Reply by sgsdxzy** [reply]
> Ok, so I just put 1girl Arzaxi, [filewords] in instance prompt, it should automatically take tags from .txt file with the same name as image, without turning on this option right? ![chrome_cn1cnONLhQ](https://user-images.githubusercontent.com/83187970/201723039-4ad821d0-51e5-49c4-bc88-6005c7daef6d.png)

Yes.

**Reply by Trauter1** [reply]
Ok, thank you for answering!

**Reply by leoftm** [reply]
https://github.com/d8ahazard/sd_dreambooth_extension/discussions/131
I know this is a multi-post since we had the exact same conversation here, but I would like to hear from someone who knows more about this.
(In filename.txt, each teacher image is set by deepdanbooru to e.g. "1girl, bracelet, earrings, long hair, solo, blue sky, blue background, sky, day, facial mark, necklace, facepaint, black hair, red eyes").

I read about it here.
https://dreambooth.github.io/DreamBooth_files/system.png 
https://dreambooth.github.io

If the concept I want to teach is, for example, "Arzaxi", the necessary settings are
The instance prompt is "Arzaxi,[filename.txt]" and
I understood that the required class prompt is "[filename.txt]".
It looks like it is also possible with instance prompt is "Arzaxi ,1girl" and class prommt is "1girl".
Is my understanding of this wrong?

**Comment by sgsdxzy** [comment]
And those two setting, they are no longer necessary as you can just use `[filewords]` as your prompt to achieve the same result. If you check them however, they override [filewords] behavior and your prompt is ignored.
They will be removed from ui once #85 is fully merged.

**Comment by Trauter1** [comment]
@sgsdxzy Started generating class images with those settings :
![chrome_jkEUXdNDBX](https://user-images.githubusercontent.com/83187970/201730288-8b184cda-7005-4697-be4f-e3b10e5e1155.png)
But looking at them, they don't resemble prompts at all.
![4-ffc2f5dd979bbd196390853c8c0bb1059974570e](https://user-images.githubusercontent.com/83187970/201730680-6d693e6b-7bce-4d05-aeb7-58b7f7c61c50.jpg)
Here is example of random prompt from .txt generated by me
![02307-465654922-1girl, 1girl, long hair, looking at viewer, smile, open mouth, bangs, white background, twintails, upper body, pink hair, pink e](https://user-images.githubusercontent.com/83187970/201730861-caa6b139-e960-4976-8bd6-1b845a2bb89a.png)
Looks like [filewords] is just being used as a token for prompt, instead of being replaced.
db_config.json looks like this:
`{"model_name": "Olxc", "scheduler": "ddim", "src": "final-pruned.ckpt [925997e9]", "total_steps": 0, "pretrained_model_name_or_path": "Olxc", "instance_data_dir": "/notebooks/Training datas/AliceEnd", "class_data_dir": "", "instance_prompt": "1girl Olxc, [filewords]", "use_filename_as_label": false, "use_txt_as_label": false, "class_prompt": "1girl, [filewords]", "save_sample_prompt": "", "save_sample_negative_prompt": "", "n_save_sample": 1, "save_guidance_scale": 7.5, "save_infer_steps": 30, "with_prior_preservation": true, "num_class_images": 1000, "output_dir": "/notebooks/stable-diffusion-webui/models/dreambooth/Olxc", "working_dir": "/notebooks/stable-diffusion-webui/models/dreambooth/Olxc/working", "resolution": 512, "center_crop": false, "train_text_encoder": true, "train_batch_size": 1, "sample_batch_size": 1, "num_train_epochs": 1, "max_train_steps": 16000, "gradient_accumulation_steps": 1, "gradient_checkpointing": true, "learning_rate": 1e-06, "scale_lr": false, "lr_scheduler": "constant", "lr_warmup_steps": 0, "use_8bit_adam": true, "adam_beta1": 0.9, "adam_beta2": 0.999, "adam_weight_decay": 0.01, "adam_epsilon": 1e-08, "max_grad_norm": 1, "save_preview_every": 15999, "save_embedding_every": 2000, "mixed_precision": "fp16", "not_cache_latents": true, "concepts_list": "", "use_cpu": false, "pad_tokens": true, "hflip": true, "prior_loss_weight": 1, "seed": null}`

Edit: now looking through your older comments. I think you didn't implement [filewords] for class images yet. If so, sorry for pinging you lol

**Reply by sgsdxzy** [reply]
It should be working. Does it generate the accompanying .txt for your class images?

**Reply by Trauter1** [reply]
No, going to reinstall the extension, I'm on paperspace notebook so maybe something got broken when updating to latest version.

**Reply by Trauter1** [reply]
Sadly after updating to latest version, my training won't start no matter what setting I use. I don't even get any error in my cmd.
#142 Seems to have the same problem as me.

**Reply by sgsdxzy** [reply]
I thinks that's related to b3f8d83d7ab23260dd7df3b733c9956875054b72 adding some (I think not correct) checks. Need to be fixed by @d8ahazard 
It's
```
if not isset(config.concepts_list) and not config.use_filename_as_label and not config.use_txt_as_label:
        msg = "No instance prompt specified."
```
I think training should start as long as config.instance_prompt is not empty, not the json file path. It's fixed in my resolving merge conflicts of #85

**Reply by Trauter1** [reply]
Yeah, using .json fixes the issue. And the class images come with txt file now and look like prompts. My bad for not updating properly
